src/charm.py

```diff
@@ -0,0 +1,319 @@
+from charms.observability_libs.v1.kubernetes_service_patch import (  # type: ignore[import]
+    KubernetesServicePatch,
+)
+from charms.prometheus_k8s.v0.prometheus_scrape import (  # type: ignore[import]
+    MetricsEndpointProvider,
+)
+from charms.sdcore_nrf.v0.fiveg_nrf import NRFRequires  # type: ignore[import]
+from jinja2 import Environment, FileSystemLoader
+from lightkube.models.core_v1 import ServicePort
+from ops.charm import CharmBase, EventBase
+from ops.main import main
+from ops.model import ActiveStatus, BlockedStatus, MaintenanceStatus, WaitingStatus
+from ops.pebble import Layer
+
+logger = logging.getLogger(__name__)
+
+PROMETHEUS_PORT = 9089
+SBI_PORT = 29518
+NGAPP_PORT = 38412
+SCTP_GRPC_PORT = 9000
+AMF_DATABASE_NAME = "sdcore_amf"
+DEFAULT_DATABASE_NAME = "free5gc"
+CONFIG_DIR_PATH = "/free5gc/config"
+CONFIG_FILE_NAME = "amfcfg.conf"
+CONFIG_TEMPLATE_DIR_PATH = "src/templates/"
+CONFIG_TEMPLATE_NAME = "amfcfg.conf.j2"
+
+
+class AMFOperatorCharm(CharmBase):
+    """Main class to describe juju event handling for the 5G AMF operator."""
+
+    def __init__(self, *args):
+        super().__init__(*args)
+        self._amf_container_name = self._amf_service_name = "amf"
+        self._amf_container = self.unit.get_container(self._amf_container_name)
+        self._nrf_requires = NRFRequires(charm=self, relation_name="fiveg_nrf")
+        self._amf_metrics_endpoint = MetricsEndpointProvider(self)
+        self._service_patcher = KubernetesServicePatch(
+            charm=self,
+            ports=[
+                ServicePort(name="prometheus-exporter", port=PROMETHEUS_PORT),
+                ServicePort(name="sbi", port=SBI_PORT),
+                ServicePort(name="ngapp", port=NGAPP_PORT, protocol="SCTP"),
+                ServicePort(name="sctp-grpc", port=SCTP_GRPC_PORT),
+            ],
+        )
+        self._amf_database = DatabaseRequires(
+            self, relation_name="amf-database", database_name=AMF_DATABASE_NAME
+        )
+        self._default_database = DatabaseRequires(
+            self, relation_name="default-database", database_name=DEFAULT_DATABASE_NAME
+        )
+
+        self.framework.observe(
+            self.on.default_database_relation_joined,
+            self._on_amf_pebble_ready,
+        )
+        self.framework.observe(
+            self.on.amf_database_relation_joined,
+            self._on_amf_pebble_ready,
+        )
+        self.framework.observe(
+            self._default_database.on.database_created, self._on_amf_pebble_ready
+        )
+        self.framework.observe(self._amf_database.on.database_created, self._on_amf_pebble_ready)
+        self.framework.observe(self.on.amf_pebble_ready, self._on_amf_pebble_ready)
+        self.framework.observe(self._nrf_requires.on.nrf_available, self._on_amf_pebble_ready)
+
+    def _on_amf_pebble_ready(
+        self,
+        event: EventBase,
```

Comment:
Is there any reason why we don't want to enumerate supported event types here?

Reply:
I got a previous comment on this `pr`, that this should be enough as we only are using the event to defer when needed.
https://github.com/canonical/sdcore-amf-operator/pull/1#discussion_r1191763192